### ISSUE TITLE
[sw] polling in ROM

### ISSUE THREAD

**Issue by tjaychen** [lowRISC/opentitan#6220]
Certain setup routines (for example sram / flash scramble key request) require the ROM to make a hardware request and poll for its completion.  Polling in ROM however, can be bad if the hardware underneath failed and the ROM just never gets out of the loop.

As a result, we should think a bit about what kind of polling-timeout scheme we want in ROM and how that should terminate in the event of a failure.

**Comment by tjaychen** [comment]
@alphan 
@moidx 
is this already handled? I figure you guys probably have something planned for it already.

**Comment by alphan** [comment]
Thanks @tjaychen. Currently we are not using an explicit polling-timeout scheme. Instead, we are relying on the watchdog which is initialized early in the boot with a safe default (TBD) [here](https://github.com/lowRISC/opentitan/blob/25be34391987724e412c8bd6efcfd81cc03d1212/sw/device/silicon_creator/mask_rom/mask_rom_start.S#L147) and reconfigured based on the life cycle state [here](https://github.com/lowRISC/opentitan/blob/25be34391987724e412c8bd6efcfd81cc03d1212/sw/device/silicon_creator/mask_rom/mask_rom.c#L101). Also, ROM does not pet the watchdog, we plan to set the timeout to a value that is long enough for proper ROM execution. 

The main reason for this approach is simplicity, i.e. no iteration counts to tune based on the clock frequency or peripheral being polled e.g. sram_ctrl, flash_ctrl, otbn, keymgr, hmac, uart, etc.

**Comment by alphan** [comment]
I think we can close this since we now can
* Attach a debugger,
* Build programs to be executed from SRAM (e.g. for dumps over UART) and inject them.

@tjaychen @dmcardle plmk if you have any concerns.

**Comment by tjaychen** [comment]
it's not entirely the same, because the assumption is that some kind of lock-up is happening after we enter life cycle PROD, which means the debugger would no longer be available.  We have seen this from time to time just because during production there's more volume and it can expose corner cases not previously thought of. 

We don't necessarily need a fall through like what i'm suggesting, but just have some kind of plan of how we should debug if things were to get stuck.

**Comment by alphan** [comment]
Yeah, prod can be tricky. I think this deserves an offline discussion.

**Comment by msfschaffner** [comment]
@timothytrippel @moidx @cfrantz @cdgori should we pick this discussion up again for PROD?

**Comment by cfrantz** [comment]
As of 0539d58b7d5db476863049d935c66f4c7bac745a (and to my best knowledge)

Known polling loops without timeouts:
- uart functions: spin waiting if TX FIFO is full.
- Some functions (none currently used in ROM) spin waiting for TX FIFO empty.

Known polling loops with timeouts:
- pwrmgr_cdc_sync spins with a CPU mcycle timeout (the timeout is the amount of time a CDC sync should take, plus a small amount of error overhead).
- shutdown's UART functions spin with a CPU mcycle timeout (the timeout is the amount of time it should take to empty the TX fifo).
